lib/tasks/upgrade.rake

```diff
@@ -0,0 +1,56 @@
+# TRANSLATORS: do not translate
+desc <<-END_DESC
+  This task runs predefined upgrade steps.
+
+  Examples:
+    rake upgrade:run
+    rake upgrade:mark_as_ran TASK_NAME=some_rake_task,another_rake_task
+
+END_DESC
+
+namespace :upgrade do
+  task :run => :environment do
+    ENV['FOREMAN_UPGRADE'] = '1'
+    tasks = Foreman::Plugin.registered_plugins.values.map{|plugin| plugin.upgrade_tasks}.flatten
+    tasks.select!{|task| task.run?}
+
+    tasks.each_with_index do |task, index|
+      count = "#{index + 1}/#{tasks.count}"
+
+      message = if task.long_running?
+                  _("Upgrade Step %{count}: %{name}, This may take a long while.")
+                else
+                  _("Upgrade Step %{count}: %{name}")
+                end
+
+      puts '============================================='
+      puts message % {:name => task.name, :count => count}
+      task.mark_as_ran if run_task(task)
+    end
+  end
+
+  def run_task(task)
+    begin
+      Rake::Task[task.rake_task_name].execute
+    rescue => e
+      puts _("Failed upgrade task: #{task.name}, see logs for more information.")
+
+      if task.skip_failure?
+        Rails.logger.error(_("Failed upgrade task: #{task.name}"))
+        Rails.logger.error(e.message)
+        Rails.logger.error(e.backtrace.join("\n"))
+        return false
+      else
+        raise e
```

Comment:
Should this raise a ForemanException ?

Reply:
That may hide some of the underlying error, not sure it would help.

Reply:
to avoid loosing e's backtrace, Foreman::Logging.exception("...", e) should be used